diff/diff.go

```diff
@@ -0,0 +1,293 @@
+package diff
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"reflect"
+	"sort"
+	"strings"
+)
+
+// Difference represents a "chunk" of difference between two
+// `Object`s.
+type Difference interface {
+	Summarise(out io.Writer)
+	AtPath() string
+}
+
+type Chunk struct {
+	Deleted []interface{}
+	Added   []interface{}
+	Path    string
+}
+
+func (c Chunk) AtPath() string {
+	return c.Path
+}
+
+// Some convenience funcs for when a single value has changed,
+// appeared or disappeared.
+
+func Changed(A, B interface{}, path string) Difference {
+	return Chunk{
+		Path:    path,
+		Deleted: []interface{}{A},
+		Added:   []interface{}{B},
+	}
+}
+
+func Added(B interface{}, path string) Difference {
+	return Chunk{
+		Path:  path,
+		Added: []interface{}{B},
+	}
+}
+
+func Removed(A interface{}, path string) Difference {
+	return Chunk{
+		Path:    path,
+		Deleted: []interface{}{A},
+	}
+}
+
+// a value has changed, but don't report the before or after
+type OpaqueChunk struct {
+	Path string
+}
+
+func (c OpaqueChunk) AtPath() string {
+	return c.Path
+}
+
+// Objects are considered unique by {Namespace, Kind, Name}
+type ObjectID struct {
+	Namespace string
+	Kind      string
+	Name      string
+}
+
+func (id ObjectID) String() string {
+	if id.Namespace == "" {
+		return id.Kind + " " + id.Name
+	}
+	return fmt.Sprintf("%s %s/%s", id.Kind, id.Namespace, id.Name)
+}
+
+type Object interface {
+	ID() ObjectID
+	Source() string
+}
+
+// ObjectSet is a set of several objects which can be diffed
+// collectively.
+type ObjectSet struct {
+	Source  string
+	Objects map[ObjectID]Object
+}
+
+func MakeObjectSet(source string) *ObjectSet {
+	return &ObjectSet{
+		Source:  source,
+		Objects: map[ObjectID]Object{},
+	}
+}
+
+type ObjectSetDiff struct {
+	A, B      *ObjectSet
+	OnlyA     []Object
```

Comment:
Why do we need OnlyA/B? Isn't that information contained within the Differences.Added?